File: benchmarks/perl/string_processing.pl
#!/usr/bin/env perl
# String processing benchmark

use strict;
use warnings;
use Time::HiRes qw(gettimeofday tv_interval);

# Process strings of various sizes
sub process_strings {
    my ($count, $size) = @_;
    my $result = "";
    
    for (my $i = 0; $i < $count; $i++) {
        my $str = create_random_string($size);
        my $processed = process_string($str);
        $result .= $processed;
    }
    
    return $result;
}

# Create a random string of given size
sub create_random_string {
    my ($size) = @_;
    my $chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    my $result = "";
    
    for (my $i = 0; $i < $size; $i++) {
        my $idx = int(rand(length($chars)));
        $result .= substr($chars, $idx, 1);
    }
    
    return $result;
}

# Process a string according to the rules
sub process_string {
    my ($input) = @_;
    my $result = $input;
    
    # Replace all vowels with their uppercase version
    $result =~ s/a/A/g;
    $result =~ s/e/E/g;
    $result =~ s/i/I/g;
    $result =~ s/o/O/g;
    $result =~ s/u/U/g;
    
    # Replace all digits with their doubled value
    for my $i (0..9) {
        my $digit = "$i";
        my $doubled = "" . ($i * 2);
        $result =~ s/$digit/$doubled/g;
    }
    
    # Capitalize the first letter
    if (length($result) > 0) {
        substr($result, 0, 1) = uc(substr($result, 0, 1));
    }
    
    # Reverse the string
    $result = reverse($result);
    
    # Take the first half of the reversed string
    my $half_len = int(length($result) / 2);
    $result = substr($result, 0, $half_len);
    
    return $result;
}

# Main
my $start_time = [gettimeofday];

# Process strings of different sizes
my $small = process_strings(10000, 10);   # 10,000 strings of length 10
my $medium = process_strings(1000, 100);  # 1,000 strings of length 100
my $large = process_strings(100, 1000);   # 100 strings of length 1,000

my $result_length = length($small) + length($medium) + length($large);
print "Processed string length: $result_length\n";

my $elapsed = tv_interval($start_time) * 1000;
printf "Time taken: %.2f ms\n", $elapsed;

# Perl doesn't have a standard way to get memory usage
print "Memory monitoring not available for Perl implementation\n";

File: benchmarks/php/string_processing.php
#!/usr/bin/env php
<?php
// String processing benchmark

function processStrings($count, $size) {
    $result = "";
    
    for ($i = 0; $i < $count; $i++) {
        $str = createRandomString($size);
        $processed = processString($str);
        $result .= $processed;
    }
    
    return $result;
}

function createRandomString($size) {
    $chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    $result = "";
    
    for ($i = 0; $i < $size; $i++) {
        $idx = rand(0, strlen($chars) - 1);
        $result .= $chars[$idx];
    }
    
    return $result;
}

function processString($input) {
    $result = $input;
    
    // Replace all vowels with their uppercase version
    $result = str_replace("a", "A", $result);
    $result = str_replace("e", "E", $result);
    $result = str_replace("i", "I", $result);
    $result = str_replace("o", "O", $result);
    $result = str_replace("u", "U", $result);
    
    // Replace all digits with their doubled value
    for ($i = 0; $i < 10; $i++) {
        $digit = (string)$i;
        $doubled = (string)($i * 2);
        $result = str_replace($digit, $doubled, $result);
    }
    
    // Capitalize the first letter
    if (strlen($result) > 0) {
        $result = ucfirst($result);
    }
    
    // Reverse the string
    $reversed = strrev($result);
    
    // Take the first half of the reversed string
    $halfLen = (int)(strlen($reversed) / 2);
    $result = substr($reversed, 0, $halfLen);
    
    return $result;
}

function main() {
    $startTime = microtime(true);
    
    // Process strings of different sizes
    $small = processStrings(10000, 10);   // 10,000 strings of length 10
    $medium = processStrings(1000, 100);  // 1,000 strings of length 100
    $large = processStrings(100, 1000);   // 100 strings of length 1,000
    
    $resultLength = strlen($small) + strlen($medium) + strlen($large);
    echo "Processed string length: $resultLength\n";
    
    $elapsed = (microtime(true) - $startTime) * 1000;
    echo "Time taken: $elapsed ms\n";
    
    // Get memory stats
    $memoryUsed = memory_get_peak_usage(true) / 1024;
    echo "Memory used: $memoryUsed KB\n";
}

main();